zebra-chain/src/serialization/zcash_deserialize.rs

```diff
@@ -49,3 +49,15 @@ impl<R: io::Read> ZcashDeserializeInto for R {
         T::zcash_deserialize(self)
     }
 }
+
+/// Blind preallocation of a Vec<T: SafePreallocate> can be done safely. This is in contrast
+/// to blind preallocation of a generic Vec<T>, which is a DOS vector.
+///
+/// The max_allocation() function provides a loose upper bound on the size of the Vec<T: SafePreallocate>
+/// which can possibly be received from an honest peer. If this limit is too low, Zebra may reject valid messages.
+/// In the worst case, setting the lower bound to low could cause Zebra to fall out of consensus by rejecting all messages containing a valid block.
+pub trait SafePreallocate {
```

Comment:
`unsafe` is a Rust keyword with a very specific meaning:
https://doc.rust-lang.org/book/ch19-01-unsafe-rust.html

Memory corruption is covered by Rust's safety guarantees, but memory denial of service isn't. So let's use "trusted" instead of "safe", to avoid confusion.
```suggestion
/// Blind preallocation of a Vec<T: TrustedPreallocate> is based on a bounded length. This is in contrast
/// to blind preallocation of a generic Vec<T>, which is a DOS vector.
///
/// The max_allocation() function provides a loose upper bound on the size of the Vec<T: SafePreallocate>
/// which can possibly be received from an honest peer. If this limit is too low, Zebra may reject valid messages.
/// In the worst case, setting the lower bound to low could cause Zebra to fall out of consensus by rejecting all messages containing a valid block.
pub trait TrustedPreallocate {
```